snuba/query/processors/uniq_in_select_and_having.py

```diff
@@ -0,0 +1,97 @@
+"""This is a hacky query processor put in for the purpose of
+mitigating clickhouse crashes. When a query is sent to clickhouse
+which has a `uniq` in the `HAVING` clause but not in the `SELECT`
+clause, the query node will crash. This may be fixed with future versions
+of clickhouse but we have not upgraded yet. This exists as a check so that we
+return an exception to the user but don't crash the clickhouse query node
+"""
+
+import logging
+from dataclasses import fields
+from typing import List, Sequence
+
+from snuba.clickhouse.processors import QueryProcessor
+from snuba.clickhouse.query import Query
+from snuba.query.exceptions import InvalidQueryException
+from snuba.query.expressions import Column, Expression, FunctionCall, NoopVisitor
+from snuba.request.request_settings import RequestSettings
+from snuba.state import get_config
+
+
+class MismatchedAggregationException(InvalidQueryException):
+    pass
+
+
+class _FunctionNameFindingVisitor(NoopVisitor):
+    def __init__(self, function_name_to_find: str):
+        self.function_name_to_find = function_name_to_find
+        self.found_functions: List[FunctionCall] = []
+
+    def visit_function_call(self, exp: FunctionCall) -> None:
+        for param in exp.parameters:
+            param.accept(self)
+        if exp.function_name == self.function_name_to_find:
+            self.found_functions.append(exp)
+        return None
+
+
+def _expressions_equal_ignore_aliases(exp1: Expression, exp2: Expression) -> bool:
+    if type(exp1) != type(exp2):
+        return False
+    fields_equal = True
+    for field in fields(exp1):
+        if field.name == "alias":
+            continue
+        field_val1 = getattr(exp1, field.name)
+        field_val2 = getattr(exp2, field.name)
+        if isinstance(field_val1, Expression):
+            fields_equal &= _expressions_equal_ignore_aliases(field_val1, field_val2)
+        else:
+            fields_equal &= field_val1 == field_val2
+        if not fields_equal:
+            return False
+    return fields_equal
+
+
+class _ExpressionOrAliasMatcher(NoopVisitor):
+    def __init__(self, expressions_to_match: Sequence[FunctionCall]):
+        self.expressions_to_match = expressions_to_match
+        self.found_expressions = [False] * len(expressions_to_match)
+
+    def visit_column(self, exp: Column) -> None:
+        for i, exp_to_match in enumerate(self.expressions_to_match):
+            if exp_to_match.alias is not None and exp_to_match.alias == exp.column_name:
+                self.found_expressions[i] = True
```

Comment:
This will never happen. You can remove this method entirely. 
The parsing operation expands all the aliases by design. After parsing alias references do not exist anymore in the query. 
That was intentional so that the query processing does not need to deal with them. We also tweak them (`_snuba_aliasname`) so that we would not have shadowing of real columns.

This means that if you had 
```
SELECT u_user
...
HAVING (uniq(user) as u_user) > x
```

In the AST, at every step of query processing you will have 
```
SELECT (uniq(user) as u_user)
....
HAVING (uniq(user) as u_user) > x
```

So you do not have to resolve aliases, you just need to be consistent with alias assignment if you apply changes that are not across the board. Changes are generally applied across the board as the transformation functions in the AST api have very little context of what they are transforming. 
I can point you to the document where we discussed multiple options for this case.

Reply:
I'm okay to remove this however I'm wondering if this breaks the modularity of the pipeline. In order to make this assumption, this stage of the pipeline has to know about the behavior of the parsing stage. Is that a good thing?

Reply:
got it